### PR TITLE
Add `prefer-t-throws` rule

### PR DESCRIPTION
**Pull request by Mesteery**
Fixes: https://github.com/avajs/eslint-plugin-ava/issues/156

### CODE REVIEW

File: docs/rules/prefer-t-throws.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+This rule will enforce the use of `t.throws()` when possible.
+
+## Fail
+
+```js
+const test = require('ava');
+
+test('some test', async t => {
+	try {
+		await Promise.reject('error');
```
Comment (sindresorhus) [inline docs/rules/prefer-t-throws.md:12]:
```suggestion
		await Promise.reject('error');
```

This will confuse users. I know it's just an example, but it makes no sense.

Comment (sindresorhus) [inline docs/rules/prefer-t-throws.md:12]:
```suggestion
		await potentiallyThrowingFunction();
```

File: docs/rules/prefer-t-throws.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+		t.fail();
+	} catch (error) {
+		t.is(error, 'error');
```
Comment (sindresorhus) [inline docs/rules/prefer-t-throws.md:15]:
```suggestion
		t.is(error.message, 'Unicorn overload');
```

File: docs/rules/prefer-t-throws.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+		await promiseThatMaybeFails();
+		await anotherPromise;
+		await timeout(100, 'good error message');
+		t.fail();
+	} catch (error) {
+		t.is(error, 'good error message');
+	}
+});
+```
+
```
Comment (sindresorhus) [inline docs/rules/prefer-t-throws.md:60]:
```suggestion
```

**Comment by sindresorhus** [comment]
AVA now has `t.throws` and `t.throwsAsync`. See the AVA docs.

File: test/prefer-t-throws.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+		`${header}test(async t => { try { await promise; } catch (error) { t.is(error, 'error'); } });`,
+		`${header}test(async t => { try { await promise; await anotherPromise; t.fail(); } catch (error) { t.is(error, 'error'); } });`,
+	],
+	invalid: [
+		{
+			code: `${header}test(async t => { try { await Promise.reject('error'); t.fail(); } catch (error) { t.is(error, 'error'); } });`,
+			errors: [{message: 'Prefer using the `t.throws()` assertion.'}],
+		},
+	],
+});
```
Comment (sindresorhus) [inline test/prefer-t-throws.js:27]: I'd like to see more tests. Try to think of edge-cases.

**Comment by Mesteery** [comment]
> AVA now has t.throws and t.throwsAsync. See the AVA docs.

So should the rule be `t-throws-async` instead or should I add another rule?

**Comment by sindresorhus** [comment]
The name should be the what it is, but you need to also support the async method.

File: rules/prefer-t-throws.js [hunk 5]
```diff
@@ -0,0 +1,146 @@
+'use strict';
+
+const {visitIf} = require('enhance-visitors');
+const createAvaRule = require('../create-ava-rule');
+const util = require('../util');
+
+function hasAwaitExpression(nodes) {
```
Comment (sindresorhus) [inline rules/prefer-t-throws.js:7]: There must be a simpler way to achieve this.

Comment (Mesteery) [reply rules/prefer-t-throws.js:7]: I don't know, but the function does a bit more than its name indicates: it looks for an AwaitExpression, which is not in a (arrow) function, and throws if it meets a return.

Comment (sindresorhus) [reply rules/prefer-t-throws.js:7]: At least add a code comments that succinctly explains what the function does.

File: docs/rules/prefer-t-throws.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+## Pass
+
+```js
+const test = require('ava');
+
+test('some test', async t => {
+	const error = await t.throwsAsync(asyncThrowingFunction());
+	t.is(error.message, 'Unicorn overload');
```
Comment (sindresorhus) [inline docs/rules/prefer-t-throws.md:55]: This is not the recommended way to test the error message.

File: rules/prefer-t-throws.js [hunk 7]
```diff
@@ -0,0 +1,149 @@
+'use strict';
+
+const {visitIf} = require('enhance-visitors');
+const createAvaRule = require('../create-ava-rule');
+const util = require('../util');
+
+// This function checks if there is an AwaitExpression, which is not inside another function.
```
Comment (sindresorhus) [inline rules/prefer-t-throws.js:7]: Maybe you could find an `AwaitExpression`, then iterate upwards through its parent scopes until you encounter a function, and then check the function.

https://eslint.org/docs/developer-guide/scope-manager-interface

File: docs/rules/prefer-t-throws.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+});
+```
+
+```js
+const test = require('ava');
+
+test('some test', async t => {
+	try {
+		const result = await potentiallyThrowingFunction();
+		// This passes because using a try-catch can be intentional, for example, to test both the result and the error:
```
[the remark this answers is not in the document]
Comment (Mesteery) [reply docs/rules/prefer-t-throws.md:65]: Indeed, I can't remember what was in my head.

File: rules/prefer-t-throws.js (hunk 7 above)
Comment (Mesteery) [reply rules/prefer-t-throws.js:7]: What about the `ReturnStatement`? I doubt there's anything simpler than that in the end.

File: docs/rules/prefer-t-throws.md (hunk 6 above)
Comment (sindresorhus) [reply docs/rules/prefer-t-throws.md:55]: That is also not the recommended way. Please read the docs for `t.throwsAsync`. It accepts a `message` option.

File: rules/prefer-t-throws.js (hunk 7 above)
Comment (sindresorhus) [reply rules/prefer-t-throws.js:7]: Do you have a realistic example of when someone would actually use `return` in a try/catch when testing an error?

Comment (Mesteery) [reply rules/prefer-t-throws.js:7]: No, I don't have one. It's pretty hard to find a realistic example but I have a feeling that there might be some code written that way.

Comment (sindresorhus) [reply rules/prefer-t-throws.js:7]: > I have a feeling that there might be some code written that way.

Rule logic should be based on test fixtures, not feelings. My recommendation to simplify the logic here stands.
